### PR TITLE
call `KernelFillGaps*` from device

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- move the source code of the functor `KernelShiftParticles` under `KernelFillGaps` 
- remove host side call of `KernelFillGaps` and `KernelFillGapsLastFrame`
- call `KernelFillGaps` on device from `KernelShiftParticles`
- call `KernelFillGapsLastFrame` on device from `KernelFillGaps`
- add documentation

# Improvements
This PR dramatically  reduces the number of kernels called per time step and therefore improves the performance (~2%) and maybe the scaling.

# Note for the Reviewer
The first commit changes no algorithms, only the source code of the functor `KernelShiftParticles` is moved after `KernelFillGaps`.
The second commit contains all important changes.

# Tests

- [x] default 3D KHI

### CODE REVIEW

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 1]
```diff
@@ -755,6 +459,324 @@ struct KernelFillGaps
...
+
+        // fill all gaps in the last frame of the supercell
+        KernelFillGapsLastFrame< numWorkers >{ }(
+            pb,
+            mapper
+        );
+    }
+};
+
+/** shift particles those leaving the supercell
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: ~~those~~

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 2]
```diff
@@ -755,6 +459,324 @@ struct KernelFillGaps
...
+ *
+ * The functor fulfills the restriction that all frames expect the last
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: expect -> except

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 3]
```diff
@@ -755,6 +459,324 @@ struct KernelFillGaps
...
+ * must be filled with as many particles as can be stored in a frame.
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: -> must be fully filled with particles.

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 4]
```diff
@@ -755,6 +459,324 @@ struct KernelFillGaps
...
+            );
+            __syncthreads( );
+
+            forEachExchange(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    /* if the `low frame` is now full, each master thread removes it and
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: I think the `low frame` is not really *removed* but just not managed any more by the master thread (and a new one taken instead), right?

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 5]
```diff
@@ -457,14 +459,28 @@ struct KernelFillGaps
...
+ * must be filled with as many particles as can be stored in a frame.
+ *
+ * @tparam T_numWorkers number of workers
+ */
 template< uint32_t T_numWorkers >
 struct KernelShiftParticles
 {
-    /*! This kernel move particles to the next supercell
-     * This kernel can only run with a double checker board
+    /** This kernel move particles to the next supercell
```
Comment (theZiz) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: move**s**

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 6]
```diff
@@ -52,226 +52,129 @@ getPreviousFrameAndRemoveLastFrame( const typename T_ParticleBox::FramePtr& fram
...
-        constexpr uint32_t numExchanges = traits::NumberOfExchanges< dim >::value;
+        constexpr uint32_t frameSize = math::CT::volume<typename T_Mapping::SuperCellSize>::type::value;
+        constexpr uint32_t dim = T_Mapping::Dim;
         constexpr uint32_t numWorkers = T_numWorkers;
 
-        /* define memory for two times Exchanges
-         * index range [0,numExchanges-1] are being referred to as `low frames`
-         * index range [numExchanges,2*numExchanges-1] are being referred to as `high frames`
-         */
+        typedef typename T_ParBox::FramePtr FramePtr;
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: ah, a *typedef* ;)

Comment (psychocoderHPC) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: `Krümelkacker` ^^

**Comment by psychocoderHPC** [comment]
updated

File: src/libPMacc/include/particles/ParticlesBase.kernel (hunk 6 above)
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: sorry ^^ but I did not enforce it xD

**Pull request merged**